# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    
    source_id_text =[]
    source_sentences = source_text.split('\n')
    for source_raw_sentence in source_sentences:
        final_sentence = source_raw_sentence.split()
        lettuce = [source_vocab_to_int[word]for word in final_sentence]
        source_id_text.append(lettuce)
    print(source_id_text)
    
    eos =[target_vocab_to_int['<EOS>']]
    target_id_text =[]
    target_sentences = target_text.split('\n')
    for target_raw_sentence in target_sentences:
        final_sentence = target_raw_sentence.split()
        lettuce = [target_vocab_to_int[word]for word in final_sentence]+eos
        target_id_text.append(lettuce)
    print(target_id_text)
    
    return source_id_text,target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

[[21, 14, 17, 31, 29, 28, 10, 18, 7, 5, 17, 15, 16, 22, 11], [4, 24, 32, 17, 19, 20, 28, 8, 18, 7, 5, 17, 19, 25, 16, 13, 11], [27, 17, 19, 29, 28, 9, 18, 7, 5, 17, 19, 12, 16, 26, 11], [4, 24, 32, 17, 31, 23, 28, 26, 18, 7, 5, 17, 30, 16, 6, 11]]
[[19, 13, 21, 16, 12, 9, 14, 28, 18, 32, 29, 21, 6, 33, 27, 10, 1], [7, 20, 21, 31, 4, 33, 5, 18, 32, 29, 24, 25, 33, 22, 10, 1], [30, 21, 31, 12, 33, 8, 18, 32, 29, 21, 31, 15, 33, 23, 10, 1], [7, 20, 21, 16, 11, 33, 23, 18, 32, 29, 17, 4, 33, 26, 10, 1]]
Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.


# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    
    input = tf.placeholder(tf.int32,[None, None],name = "input")
    targets = tf.placeholder(tf.int32,[None, None],name = "targets")
    learning_rate = tf.placeholder(tf.float32,name = "learning_rate")
    keep_prob = tf.placeholder(tf.float32,name = "keep_prob")
    target_sequence_length = tf.placeholder(tf.int32, (None,), name = "target_sequence_length")
    max_target_len = tf.reduce_max(target_sequence_length)
    source_sequence_length = tf.placeholder(tf.int32,(None,),name = "source_sequence_length")
    
    return input, targets, learning_rate, keep_prob, target_sequence_length, max_target_len, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    preprocessed_target_data = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    return preprocessed_target_data
    
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # Encoder embedding
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)

    def build_cell(rnn_size, keep_prob):
        lstm = tf.contrib.rnn.BasicLSTMCell(rnn_size)
        drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
        return drop

    enc_cell_with_dropout=tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    enc_output, enc_state = tf.nn.dynamic_rnn(enc_cell_with_dropout, enc_embed_input, sequence_length=source_sequence_length, dtype=tf.float32)
    return enc_output, enc_state
    
    

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    training_helper = tf.contrib.seq2seq.TrainingHelper(dec_embed_input,target_sequence_length, name="training_helper")
    seq2seq_basic_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,training_helper,encoder_state,output_layer)
    BasicDecoderOutput,_ = tf.contrib.seq2seq.dynamic_decode(seq2seq_basic_decoder,impute_finished=True,maximum_iterations=max_summary_length)
    return BasicDecoderOutput



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')
    
    emb_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,start_tokens,end_of_sequence_id)
    seq2seq_basic_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,emb_helper,encoder_state,output_layer)
    result,_ = tf.contrib.seq2seq.dynamic_decode(seq2seq_basic_decoder,impute_finished=True,maximum_iterations=max_target_sequence_length)
    return result



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size,initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return dec_cell
    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
 
    output_layer = Dense(target_vocab_size,kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))

    with tf.variable_scope("decode") as decoding_scope:
        training_decoder_output = decoding_layer_train(encoder_state, dec_cell, dec_embed_input,target_sequence_length, max_target_sequence_length, output_layer, keep_prob)

    with tf.variable_scope('decode', reuse=True) as decoding_scope:
        inference_decoder_output = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, target_vocab_to_int['<GO>'],target_vocab_to_int['<EOS>'], max_target_sequence_length,target_vocab_size, output_layer, batch_size, keep_prob)

    return training_decoder_output, inference_decoder_output
    
    
    
    

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    enc_input, enc_state = encoding_layer(input_data,rnn_size,num_layers,keep_prob,source_sequence_length,source_vocab_size,enc_embedding_size)
    preprocessed_target_data = process_decoder_input(target_data,target_vocab_to_int,batch_size)
    training_basic_decoder_output,inference_basic_decoder_output = decoding_layer(target_data,enc_state,target_sequence_length, max_target_sentence_length,rnn_size,num_layers,target_vocab_to_int,target_vocab_size,batch_size, keep_prob, dec_embedding_size)
    return training_basic_decoder_output,inference_basic_decoder_output 


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [14]:
# Number of Epochs
epochs = 1
# Batch Size
batch_size = 10
# RNN Size
rnn_size = 2
num_layers = 2
# Embedding Size
encoding_embedding_size = 15
decoding_embedding_size = 15
# Learning Rate
learning_rate = 0.02
# Dropout Keep Probability
keep_probability = 0.3
display_step = 5

### Build the Graph
Build the graph using the neural network you implemented.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    5/13786 - Train Accuracy: 0.1778, Validation Accuracy: 0.0529, Loss: 5.7015
Epoch   0 Batch   10/13786 - Train Accuracy: 0.0278, Validation Accuracy: 0.0353, Loss: 5.4205
Epoch   0 Batch   15/13786 - Train Accuracy: 0.0333, Validation Accuracy: 0.0353, Loss: 5.0914
Epoch   0 Batch   20/13786 - Train Accuracy: 0.3150, Validation Accuracy: 0.0529, Loss: 4.7502
Epoch   0 Batch   25/13786 - Train Accuracy: 0.1789, Validation Accuracy: 0.0529, Loss: 4.4567
Epoch   0 Batch   30/13786 - Train Accuracy: 0.1263, Validation Accuracy: 0.0529, Loss: 4.3964
Epoch   0 Batch   35/13786 - Train Accuracy: 0.2158, Validation Accuracy: 0.0529, Loss: 4.1541
Epoch   0 Batch   40/13786 - Train Accuracy: 0.2333, Validation Accuracy: 0.0529, Loss: 4.0914
Epoch   0 Batch   45/13786 - Train Accuracy: 0.1000, Validation Accuracy: 0.0529, Loss: 4.0515
Epoch   0 Batch   50/13786 - Train Accuracy: 0.1056, Validation Accuracy: 0.0529, Loss: 4.0835
Epoch   0 Batch   55/13786 - Train Accuracy: 0.272

Epoch   0 Batch  440/13786 - Train Accuracy: 0.1000, Validation Accuracy: 0.1059, Loss: 2.6996
Epoch   0 Batch  445/13786 - Train Accuracy: 0.0778, Validation Accuracy: 0.1059, Loss: 2.5747
Epoch   0 Batch  450/13786 - Train Accuracy: 0.0850, Validation Accuracy: 0.1059, Loss: 2.6036
Epoch   0 Batch  455/13786 - Train Accuracy: 0.0778, Validation Accuracy: 0.1059, Loss: 2.7737
Epoch   0 Batch  460/13786 - Train Accuracy: 0.0833, Validation Accuracy: 0.1059, Loss: 2.6308
Epoch   0 Batch  465/13786 - Train Accuracy: 0.0944, Validation Accuracy: 0.1059, Loss: 2.7652
Epoch   0 Batch  470/13786 - Train Accuracy: 0.0882, Validation Accuracy: 0.1059, Loss: 2.6634
Epoch   0 Batch  475/13786 - Train Accuracy: 0.0650, Validation Accuracy: 0.1059, Loss: 2.2958
Epoch   0 Batch  480/13786 - Train Accuracy: 0.0722, Validation Accuracy: 0.1059, Loss: 2.6688
Epoch   0 Batch  485/13786 - Train Accuracy: 0.0722, Validation Accuracy: 0.1059, Loss: 2.6558
Epoch   0 Batch  490/13786 - Train Accuracy: 0.088

Epoch   0 Batch  875/13786 - Train Accuracy: 0.1167, Validation Accuracy: 0.1059, Loss: 2.3167
Epoch   0 Batch  880/13786 - Train Accuracy: 0.0632, Validation Accuracy: 0.0941, Loss: 2.2941
Epoch   0 Batch  885/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0235, Loss: 2.1129
Epoch   0 Batch  890/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0235, Loss: 2.2085
Epoch   0 Batch  895/13786 - Train Accuracy: 0.0176, Validation Accuracy: 0.0235, Loss: 2.5920
Epoch   0 Batch  900/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0235, Loss: 2.3252
Epoch   0 Batch  905/13786 - Train Accuracy: 0.0250, Validation Accuracy: 0.0235, Loss: 2.1247
Epoch   0 Batch  910/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0235, Loss: 2.3930
Epoch   0 Batch  915/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0235, Loss: 2.0610
Epoch   0 Batch  920/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0235, Loss: 2.2851
Epoch   0 Batch  925/13786 - Train Accuracy: 0.022

Epoch   0 Batch 1310/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0235, Loss: 2.0969
Epoch   0 Batch 1315/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0235, Loss: 1.9408
Epoch   0 Batch 1320/13786 - Train Accuracy: 0.0368, Validation Accuracy: 0.0471, Loss: 2.1027
Epoch   0 Batch 1325/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0235, Loss: 1.9562
Epoch   0 Batch 1330/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0235, Loss: 2.0161
Epoch   0 Batch 1335/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0235, Loss: 2.0125
Epoch   0 Batch 1340/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0235, Loss: 1.8921
Epoch   0 Batch 1345/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0235, Loss: 1.8638
Epoch   0 Batch 1350/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0235, Loss: 1.9100
Epoch   0 Batch 1355/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0235, Loss: 1.9074
Epoch   0 Batch 1360/13786 - Train Accuracy: 0.010

Epoch   0 Batch 1745/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0059, Loss: 1.8197
Epoch   0 Batch 1750/13786 - Train Accuracy: 0.0316, Validation Accuracy: 0.0235, Loss: 1.7571
Epoch   0 Batch 1755/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0059, Loss: 1.8638
Epoch   0 Batch 1760/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0176, Loss: 1.8526
Epoch   0 Batch 1765/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.7654
Epoch   0 Batch 1770/13786 - Train Accuracy: 0.0250, Validation Accuracy: 0.0059, Loss: 1.6549
Epoch   0 Batch 1775/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0235, Loss: 1.7434
Epoch   0 Batch 1780/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.7909
Epoch   0 Batch 1785/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0059, Loss: 1.8524
Epoch   0 Batch 1790/13786 - Train Accuracy: 0.0118, Validation Accuracy: 0.0059, Loss: 1.8880
Epoch   0 Batch 1795/13786 - Train Accuracy: 0.010

Epoch   0 Batch 2180/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0118, Loss: 1.6962
Epoch   0 Batch 2185/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0059, Loss: 1.8049
Epoch   0 Batch 2190/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0059, Loss: 1.9127
Epoch   0 Batch 2195/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0118, Loss: 1.6104
Epoch   0 Batch 2200/13786 - Train Accuracy: 0.0333, Validation Accuracy: 0.0118, Loss: 1.8054
Epoch   0 Batch 2205/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0118, Loss: 1.7446
Epoch   0 Batch 2210/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 1.7348
Epoch   0 Batch 2215/13786 - Train Accuracy: 0.0368, Validation Accuracy: 0.0118, Loss: 1.7382
Epoch   0 Batch 2220/13786 - Train Accuracy: 0.0143, Validation Accuracy: 0.0059, Loss: 1.6671
Epoch   0 Batch 2225/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.7982
Epoch   0 Batch 2230/13786 - Train Accuracy: 0.036

Epoch   0 Batch 2615/13786 - Train Accuracy: 0.0211, Validation Accuracy: 0.0118, Loss: 1.7352
Epoch   0 Batch 2620/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0235, Loss: 1.7762
Epoch   0 Batch 2625/13786 - Train Accuracy: 0.0294, Validation Accuracy: 0.0118, Loss: 1.7464
Epoch   0 Batch 2630/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.6461
Epoch   0 Batch 2635/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0118, Loss: 1.5343
Epoch   0 Batch 2640/13786 - Train Accuracy: 0.0100, Validation Accuracy: 0.0118, Loss: 1.4355
Epoch   0 Batch 2645/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 1.6555
Epoch   0 Batch 2650/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0118, Loss: 1.7258
Epoch   0 Batch 2655/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 1.6808
Epoch   0 Batch 2660/13786 - Train Accuracy: 0.0100, Validation Accuracy: 0.0118, Loss: 1.6498
Epoch   0 Batch 2665/13786 - Train Accuracy: 0.027

Epoch   0 Batch 3050/13786 - Train Accuracy: 0.0211, Validation Accuracy: 0.0235, Loss: 1.5440
Epoch   0 Batch 3055/13786 - Train Accuracy: 0.0211, Validation Accuracy: 0.0353, Loss: 1.5869
Epoch   0 Batch 3060/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0235, Loss: 1.5907
Epoch   0 Batch 3065/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0118, Loss: 1.6426
Epoch   0 Batch 3070/13786 - Train Accuracy: 0.0421, Validation Accuracy: 0.0235, Loss: 1.6562
Epoch   0 Batch 3075/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0235, Loss: 1.4072
Epoch   0 Batch 3080/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0235, Loss: 1.4432
Epoch   0 Batch 3085/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0353, Loss: 1.7170
Epoch   0 Batch 3090/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0235, Loss: 1.6853
Epoch   0 Batch 3095/13786 - Train Accuracy: 0.0176, Validation Accuracy: 0.0235, Loss: 1.7360
Epoch   0 Batch 3100/13786 - Train Accuracy: 0.011

Epoch   0 Batch 3485/13786 - Train Accuracy: 0.0211, Validation Accuracy: 0.0118, Loss: 1.5506
Epoch   0 Batch 3490/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 1.6123
Epoch   0 Batch 3495/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0118, Loss: 1.7487
Epoch   0 Batch 3500/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.6582
Epoch   0 Batch 3505/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0118, Loss: 1.5689
Epoch   0 Batch 3510/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.6465
Epoch   0 Batch 3515/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0118, Loss: 1.4869
Epoch   0 Batch 3520/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 1.6688
Epoch   0 Batch 3525/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0118, Loss: 1.5122
Epoch   0 Batch 3530/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.4978
Epoch   0 Batch 3535/13786 - Train Accuracy: 0.011

Epoch   0 Batch 3920/13786 - Train Accuracy: 0.0316, Validation Accuracy: 0.0294, Loss: 1.6375
Epoch   0 Batch 3925/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0235, Loss: 1.3584
Epoch   0 Batch 3930/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0235, Loss: 1.4600
Epoch   0 Batch 3935/13786 - Train Accuracy: 0.0100, Validation Accuracy: 0.0353, Loss: 1.4782
Epoch   0 Batch 3940/13786 - Train Accuracy: 0.0200, Validation Accuracy: 0.0353, Loss: 1.3052
Epoch   0 Batch 3945/13786 - Train Accuracy: 0.0316, Validation Accuracy: 0.0353, Loss: 1.5697
Epoch   0 Batch 3950/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0235, Loss: 1.6960
Epoch   0 Batch 3955/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0235, Loss: 1.3683
Epoch   0 Batch 3960/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0235, Loss: 1.7012
Epoch   0 Batch 3965/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0235, Loss: 1.8140
Epoch   0 Batch 3970/13786 - Train Accuracy: 0.010

Epoch   0 Batch 4355/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0235, Loss: 1.3470
Epoch   0 Batch 4360/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.4218
Epoch   0 Batch 4365/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0235, Loss: 1.3783
Epoch   0 Batch 4370/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0235, Loss: 1.4480
Epoch   0 Batch 4375/13786 - Train Accuracy: 0.0211, Validation Accuracy: 0.0118, Loss: 1.5844
Epoch   0 Batch 4380/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 1.5000
Epoch   0 Batch 4385/13786 - Train Accuracy: 0.0294, Validation Accuracy: 0.0118, Loss: 1.4339
Epoch   0 Batch 4390/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0235, Loss: 1.4437
Epoch   0 Batch 4395/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0176, Loss: 1.5680
Epoch   0 Batch 4400/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0235, Loss: 1.2461
Epoch   0 Batch 4405/13786 - Train Accuracy: 0.005

Epoch   0 Batch 4790/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.4687
Epoch   0 Batch 4795/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.5714
Epoch   0 Batch 4800/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.4087
Epoch   0 Batch 4805/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0235, Loss: 1.4154
Epoch   0 Batch 4810/13786 - Train Accuracy: 0.0100, Validation Accuracy: 0.0000, Loss: 1.3976
Epoch   0 Batch 4815/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0118, Loss: 1.3950
Epoch   0 Batch 4820/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0235, Loss: 1.4568
Epoch   0 Batch 4825/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0118, Loss: 1.5165
Epoch   0 Batch 4830/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0235, Loss: 1.4765
Epoch   0 Batch 4835/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.3310
Epoch   0 Batch 4840/13786 - Train Accuracy: 0.000

Epoch   0 Batch 5225/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0353, Loss: 1.2737
Epoch   0 Batch 5230/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0353, Loss: 1.4294
Epoch   0 Batch 5235/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0118, Loss: 1.3053
Epoch   0 Batch 5240/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.4033
Epoch   0 Batch 5245/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0118, Loss: 1.3798
Epoch   0 Batch 5250/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.4158
Epoch   0 Batch 5255/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 1.4211
Epoch   0 Batch 5260/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 1.5384
Epoch   0 Batch 5265/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0118, Loss: 1.4674
Epoch   0 Batch 5270/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0235, Loss: 1.2946
Epoch   0 Batch 5275/13786 - Train Accuracy: 0.005

Epoch   0 Batch 5660/13786 - Train Accuracy: 0.0316, Validation Accuracy: 0.0235, Loss: 1.5467
Epoch   0 Batch 5665/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0235, Loss: 1.1855
Epoch   0 Batch 5670/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0235, Loss: 1.4764
Epoch   0 Batch 5675/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0353, Loss: 1.2779
Epoch   0 Batch 5680/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0235, Loss: 1.2888
Epoch   0 Batch 5685/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0118, Loss: 1.4103
Epoch   0 Batch 5690/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0235, Loss: 1.4294
Epoch   0 Batch 5695/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.4174
Epoch   0 Batch 5700/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.4291
Epoch   0 Batch 5705/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0118, Loss: 1.4161
Epoch   0 Batch 5710/13786 - Train Accuracy: 0.005

Epoch   0 Batch 6095/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0353, Loss: 1.2363
Epoch   0 Batch 6100/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0353, Loss: 1.3017
Epoch   0 Batch 6105/13786 - Train Accuracy: 0.0211, Validation Accuracy: 0.0353, Loss: 1.4157
Epoch   0 Batch 6110/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0353, Loss: 1.3931
Epoch   0 Batch 6115/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0353, Loss: 1.3507
Epoch   0 Batch 6120/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0235, Loss: 1.2907
Epoch   0 Batch 6125/13786 - Train Accuracy: 0.0048, Validation Accuracy: 0.0353, Loss: 1.1438
Epoch   0 Batch 6130/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0353, Loss: 1.5270
Epoch   0 Batch 6135/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0353, Loss: 1.4184
Epoch   0 Batch 6140/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0353, Loss: 1.2339
Epoch   0 Batch 6145/13786 - Train Accuracy: 0.016

Epoch   0 Batch 6530/13786 - Train Accuracy: 0.0150, Validation Accuracy: 0.0353, Loss: 1.1394
Epoch   0 Batch 6535/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0353, Loss: 1.1602
Epoch   0 Batch 6540/13786 - Train Accuracy: 0.0211, Validation Accuracy: 0.0353, Loss: 1.2930
Epoch   0 Batch 6545/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0353, Loss: 1.2792
Epoch   0 Batch 6550/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0353, Loss: 1.2866
Epoch   0 Batch 6555/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0235, Loss: 1.2555
Epoch   0 Batch 6560/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0235, Loss: 1.2921
Epoch   0 Batch 6565/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0353, Loss: 1.0923
Epoch   0 Batch 6570/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0235, Loss: 1.2332
Epoch   0 Batch 6575/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0235, Loss: 1.4219
Epoch   0 Batch 6580/13786 - Train Accuracy: 0.017

Epoch   0 Batch 6965/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.1855
Epoch   0 Batch 6970/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 1.2703
Epoch   0 Batch 6975/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.1199
Epoch   0 Batch 6980/13786 - Train Accuracy: 0.0250, Validation Accuracy: 0.0118, Loss: 1.2258
Epoch   0 Batch 6985/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0118, Loss: 1.2441
Epoch   0 Batch 6990/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.1425
Epoch   0 Batch 6995/13786 - Train Accuracy: 0.0150, Validation Accuracy: 0.0118, Loss: 1.1655
Epoch   0 Batch 7000/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0118, Loss: 1.2522
Epoch   0 Batch 7005/13786 - Train Accuracy: 0.0150, Validation Accuracy: 0.0118, Loss: 1.2441
Epoch   0 Batch 7010/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.3164
Epoch   0 Batch 7015/13786 - Train Accuracy: 0.000

Epoch   0 Batch 7400/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 1.0918
Epoch   0 Batch 7405/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0118, Loss: 1.1478
Epoch   0 Batch 7410/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.2950
Epoch   0 Batch 7415/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0118, Loss: 1.0952
Epoch   0 Batch 7420/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 1.2524
Epoch   0 Batch 7425/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 1.2047
Epoch   0 Batch 7430/13786 - Train Accuracy: 0.0100, Validation Accuracy: 0.0118, Loss: 1.1239
Epoch   0 Batch 7435/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0118, Loss: 1.1338
Epoch   0 Batch 7440/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.1405
Epoch   0 Batch 7445/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0118, Loss: 1.3505
Epoch   0 Batch 7450/13786 - Train Accuracy: 0.011

Epoch   0 Batch 7835/13786 - Train Accuracy: 0.0250, Validation Accuracy: 0.0235, Loss: 1.0727
Epoch   0 Batch 7840/13786 - Train Accuracy: 0.0333, Validation Accuracy: 0.0235, Loss: 1.1506
Epoch   0 Batch 7845/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0235, Loss: 1.1302
Epoch   0 Batch 7850/13786 - Train Accuracy: 0.0278, Validation Accuracy: 0.0235, Loss: 1.1991
Epoch   0 Batch 7855/13786 - Train Accuracy: 0.0211, Validation Accuracy: 0.0118, Loss: 1.1579
Epoch   0 Batch 7860/13786 - Train Accuracy: 0.0316, Validation Accuracy: 0.0235, Loss: 1.1555
Epoch   0 Batch 7865/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0235, Loss: 1.1634
Epoch   0 Batch 7870/13786 - Train Accuracy: 0.0200, Validation Accuracy: 0.0118, Loss: 1.0225
Epoch   0 Batch 7875/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.2293
Epoch   0 Batch 7880/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0118, Loss: 1.2183
Epoch   0 Batch 7885/13786 - Train Accuracy: 0.011

Epoch   0 Batch 8270/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0118, Loss: 1.1435
Epoch   0 Batch 8275/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0118, Loss: 1.3673
Epoch   0 Batch 8280/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0118, Loss: 1.2223
Epoch   0 Batch 8285/13786 - Train Accuracy: 0.0211, Validation Accuracy: 0.0118, Loss: 1.1146
Epoch   0 Batch 8290/13786 - Train Accuracy: 0.0333, Validation Accuracy: 0.0118, Loss: 1.1160
Epoch   0 Batch 8295/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0235, Loss: 1.1721
Epoch   0 Batch 8300/13786 - Train Accuracy: 0.0278, Validation Accuracy: 0.0235, Loss: 1.0993
Epoch   0 Batch 8305/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 1.1313
Epoch   0 Batch 8310/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0118, Loss: 1.0916
Epoch   0 Batch 8315/13786 - Train Accuracy: 0.0100, Validation Accuracy: 0.0118, Loss: 1.1966
Epoch   0 Batch 8320/13786 - Train Accuracy: 0.000

Epoch   0 Batch 8705/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0118, Loss: 0.9894
Epoch   0 Batch 8710/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 1.0398
Epoch   0 Batch 8715/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0118, Loss: 1.1377
Epoch   0 Batch 8720/13786 - Train Accuracy: 0.0200, Validation Accuracy: 0.0118, Loss: 1.0657
Epoch   0 Batch 8725/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0118, Loss: 1.0554
Epoch   0 Batch 8730/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.2400
Epoch   0 Batch 8735/13786 - Train Accuracy: 0.0059, Validation Accuracy: 0.0235, Loss: 1.1437
Epoch   0 Batch 8740/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0118, Loss: 1.0064
Epoch   0 Batch 8745/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0118, Loss: 1.0290
Epoch   0 Batch 8750/13786 - Train Accuracy: 0.0100, Validation Accuracy: 0.0118, Loss: 1.0175
Epoch   0 Batch 8755/13786 - Train Accuracy: 0.000

Epoch   0 Batch 9140/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0118, Loss: 1.1649
Epoch   0 Batch 9145/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.0377
Epoch   0 Batch 9150/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0118, Loss: 1.0557
Epoch   0 Batch 9155/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0118, Loss: 1.0876
Epoch   0 Batch 9160/13786 - Train Accuracy: 0.0100, Validation Accuracy: 0.0118, Loss: 1.0384
Epoch   0 Batch 9165/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.1174
Epoch   0 Batch 9170/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.0497
Epoch   0 Batch 9175/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.0702
Epoch   0 Batch 9180/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 1.0732
Epoch   0 Batch 9185/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0118, Loss: 1.0388
Epoch   0 Batch 9190/13786 - Train Accuracy: 0.016

Epoch   0 Batch 9575/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 1.0258
Epoch   0 Batch 9580/13786 - Train Accuracy: 0.0200, Validation Accuracy: 0.0118, Loss: 0.9694
Epoch   0 Batch 9585/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.0788
Epoch   0 Batch 9590/13786 - Train Accuracy: 0.0211, Validation Accuracy: 0.0118, Loss: 0.9102
Epoch   0 Batch 9595/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.0415
Epoch   0 Batch 9600/13786 - Train Accuracy: 0.0059, Validation Accuracy: 0.0118, Loss: 1.2770
Epoch   0 Batch 9605/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 0.9970
Epoch   0 Batch 9610/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.0336
Epoch   0 Batch 9615/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 1.2403
Epoch   0 Batch 9620/13786 - Train Accuracy: 0.0278, Validation Accuracy: 0.0118, Loss: 1.0730
Epoch   0 Batch 9625/13786 - Train Accuracy: 0.000

Epoch   0 Batch 10010/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0118, Loss: 0.9266
Epoch   0 Batch 10015/13786 - Train Accuracy: 0.0143, Validation Accuracy: 0.0118, Loss: 0.9273
Epoch   0 Batch 10020/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0118, Loss: 0.9629
Epoch   0 Batch 10025/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.0933
Epoch   0 Batch 10030/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0118, Loss: 0.9127
Epoch   0 Batch 10035/13786 - Train Accuracy: 0.0095, Validation Accuracy: 0.0118, Loss: 1.1114
Epoch   0 Batch 10040/13786 - Train Accuracy: 0.0095, Validation Accuracy: 0.0118, Loss: 0.8787
Epoch   0 Batch 10045/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 1.2131
Epoch   0 Batch 10050/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 1.0418
Epoch   0 Batch 10055/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0118, Loss: 1.0354
Epoch   0 Batch 10060/13786 - Train Accu

Epoch   0 Batch 10440/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0118, Loss: 1.0258
Epoch   0 Batch 10445/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0118, Loss: 1.1673
Epoch   0 Batch 10450/13786 - Train Accuracy: 0.0389, Validation Accuracy: 0.0118, Loss: 0.9751
Epoch   0 Batch 10455/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0000, Loss: 1.0474
Epoch   0 Batch 10460/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 1.0506
Epoch   0 Batch 10465/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0000, Loss: 1.0547
Epoch   0 Batch 10470/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0118, Loss: 1.1253
Epoch   0 Batch 10475/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0000, Loss: 1.1997
Epoch   0 Batch 10480/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0000, Loss: 1.1428
Epoch   0 Batch 10485/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 0.9893
Epoch   0 Batch 10490/13786 - Train Accu

Epoch   0 Batch 10870/13786 - Train Accuracy: 0.0211, Validation Accuracy: 0.0118, Loss: 0.9889
Epoch   0 Batch 10875/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 0.9848
Epoch   0 Batch 10880/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.0352
Epoch   0 Batch 10885/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0118, Loss: 0.8834
Epoch   0 Batch 10890/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 0.9384
Epoch   0 Batch 10895/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0000, Loss: 0.9441
Epoch   0 Batch 10900/13786 - Train Accuracy: 0.0118, Validation Accuracy: 0.0118, Loss: 1.1633
Epoch   0 Batch 10905/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0000, Loss: 0.9362
Epoch   0 Batch 10910/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0000, Loss: 0.8396
Epoch   0 Batch 10915/13786 - Train Accuracy: 0.0100, Validation Accuracy: 0.0000, Loss: 0.9864
Epoch   0 Batch 10920/13786 - Train Accu

Epoch   0 Batch 11300/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 0.9137
Epoch   0 Batch 11305/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0059, Loss: 0.9317
Epoch   0 Batch 11310/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0000, Loss: 0.9540
Epoch   0 Batch 11315/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0000, Loss: 0.9775
Epoch   0 Batch 11320/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0059, Loss: 1.2442
Epoch   0 Batch 11325/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0000, Loss: 1.0777
Epoch   0 Batch 11330/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0118, Loss: 0.9053
Epoch   0 Batch 11335/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0118, Loss: 0.9890
Epoch   0 Batch 11340/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 0.8360
Epoch   0 Batch 11345/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0118, Loss: 1.0792
Epoch   0 Batch 11350/13786 - Train Accu

Epoch   0 Batch 11730/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0118, Loss: 1.0163
Epoch   0 Batch 11735/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0000, Loss: 0.8848
Epoch   0 Batch 11740/13786 - Train Accuracy: 0.0150, Validation Accuracy: 0.0000, Loss: 0.8839
Epoch   0 Batch 11745/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0059, Loss: 1.0290
Epoch   0 Batch 11750/13786 - Train Accuracy: 0.0150, Validation Accuracy: 0.0059, Loss: 1.0402
Epoch   0 Batch 11755/13786 - Train Accuracy: 0.0263, Validation Accuracy: 0.0000, Loss: 0.8398
Epoch   0 Batch 11760/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0059, Loss: 0.8777
Epoch   0 Batch 11765/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0059, Loss: 0.9423
Epoch   0 Batch 11770/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0059, Loss: 0.9642
Epoch   0 Batch 11775/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0059, Loss: 0.8606
Epoch   0 Batch 11780/13786 - Train Accu

Epoch   0 Batch 12160/13786 - Train Accuracy: 0.0048, Validation Accuracy: 0.0059, Loss: 0.8492
Epoch   0 Batch 12165/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0059, Loss: 0.9758
Epoch   0 Batch 12170/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0059, Loss: 0.8589
Epoch   0 Batch 12175/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0059, Loss: 0.9476
Epoch   0 Batch 12180/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0059, Loss: 0.8742
Epoch   0 Batch 12185/13786 - Train Accuracy: 0.0100, Validation Accuracy: 0.0059, Loss: 0.8673
Epoch   0 Batch 12190/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0059, Loss: 0.8834
Epoch   0 Batch 12195/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0059, Loss: 1.1442
Epoch   0 Batch 12200/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 0.8498
Epoch   0 Batch 12205/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 0.8278
Epoch   0 Batch 12210/13786 - Train Accu

Epoch   0 Batch 12590/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0059, Loss: 0.8632
Epoch   0 Batch 12595/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 1.0295
Epoch   0 Batch 12600/13786 - Train Accuracy: 0.0118, Validation Accuracy: 0.0118, Loss: 1.0738
Epoch   0 Batch 12605/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 0.8857
Epoch   0 Batch 12610/13786 - Train Accuracy: 0.0167, Validation Accuracy: 0.0059, Loss: 0.9941
Epoch   0 Batch 12615/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0059, Loss: 1.0122
Epoch   0 Batch 12620/13786 - Train Accuracy: 0.0059, Validation Accuracy: 0.0059, Loss: 0.9581
Epoch   0 Batch 12625/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0059, Loss: 0.7911
Epoch   0 Batch 12630/13786 - Train Accuracy: 0.0100, Validation Accuracy: 0.0118, Loss: 0.9087
Epoch   0 Batch 12635/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 0.8542
Epoch   0 Batch 12640/13786 - Train Accu

Epoch   0 Batch 13020/13786 - Train Accuracy: 0.0222, Validation Accuracy: 0.0118, Loss: 0.8601
Epoch   0 Batch 13025/13786 - Train Accuracy: 0.0150, Validation Accuracy: 0.0118, Loss: 0.7893
Epoch   0 Batch 13030/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 0.9688
Epoch   0 Batch 13035/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 0.8253
Epoch   0 Batch 13040/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 0.9646
Epoch   0 Batch 13045/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 0.8007
Epoch   0 Batch 13050/13786 - Train Accuracy: 0.0150, Validation Accuracy: 0.0118, Loss: 0.8479
Epoch   0 Batch 13055/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 0.8744
Epoch   0 Batch 13060/13786 - Train Accuracy: 0.0056, Validation Accuracy: 0.0118, Loss: 0.8902
Epoch   0 Batch 13065/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 0.8016
Epoch   0 Batch 13070/13786 - Train Accu

Epoch   0 Batch 13450/13786 - Train Accuracy: 0.0158, Validation Accuracy: 0.0059, Loss: 0.8901
Epoch   0 Batch 13455/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0059, Loss: 1.0802
Epoch   0 Batch 13460/13786 - Train Accuracy: 0.0111, Validation Accuracy: 0.0118, Loss: 1.0124
Epoch   0 Batch 13465/13786 - Train Accuracy: 0.0095, Validation Accuracy: 0.0176, Loss: 0.8926
Epoch   0 Batch 13470/13786 - Train Accuracy: 0.0105, Validation Accuracy: 0.0118, Loss: 0.9544
Epoch   0 Batch 13475/13786 - Train Accuracy: 0.0050, Validation Accuracy: 0.0118, Loss: 0.8078
Epoch   0 Batch 13480/13786 - Train Accuracy: 0.0053, Validation Accuracy: 0.0118, Loss: 0.8345
Epoch   0 Batch 13485/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 0.9532
Epoch   0 Batch 13490/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0059, Loss: 0.9006
Epoch   0 Batch 13495/13786 - Train Accuracy: 0.0000, Validation Accuracy: 0.0118, Loss: 0.9705
Epoch   0 Batch 13500/13786 - Train Accu

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [22]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [23]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [24]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    
    list_of_words_id = []
    lowercase_sentence = sentence.lower()
    prepared_sentence = lowercase_sentence.split()
    substitute = [vocab_to_int["<UNK>"]]
    for word in prepared_sentence:
        if word in list(vocab_to_int.keys()):
            list_of_words_id = list_of_words_id  + [vocab_to_int[word]]
        if word not in list(vocab_to_int.keys()):
            list_of_words_id = list_of_words_id  + substitute    
    return list_of_words_id


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [21]:
translate_sentence = 'new jersey is sometimes quiet during autumn , and it is snowy in april .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [85, 18, 198, 48, 45, 100, 128, 81, 66, 58, 198, 193, 138, 208, 129]
  English Words: ['new', 'jersey', 'is', 'sometimes', 'quiet', 'during', 'autumn', ',', 'and', 'it', 'is', 'snowy', 'in', 'april', '.']

Prediction
  Word Ids:      [254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254]
  French Words: votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre votre


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.